metrics/adapters/factory.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapters
+
+import (
+	"github.com/uber/jaeger-lib/metrics"
+)
+
+// FactoryWithTags creates metrics with fully qualified name and tags.
+type FactoryWithTags interface {
+	Counter(name string, tags map[string]string) metrics.Counter
+	Gauge(name string, tags map[string]string) metrics.Gauge
+	Timer(name string, tags map[string]string) metrics.Timer
+}
+
+// Options affect how the adapter factory behaves.
+type Options struct {
+	ScopeSep string
+	TagsSep  string
+	TagKVSep string
+}
+
+func defaultOptions(options Options) Options {
+	o := options
+	if o.ScopeSep == "" {
+		o.ScopeSep = "."
+	}
+	if o.TagsSep == "" {
+		o.TagsSep = "."
+	}
+	if o.TagKVSep == "" {
+		o.TagKVSep = "_"
+	}
+	return o
+}
+
+// WrapFactoryWithTags creates a real metrics.Factory that supports subscopes.
+func WrapFactoryWithTags(f FactoryWithTags, options Options) metrics.Factory {
+	return &factory{
+		Options: defaultOptions(options),
+		factory: f,
+		cache:   newCache(),
+	}
+}
+
+type factory struct {
+	Options
+	factory FactoryWithTags
+	scope   string
+	tags    map[string]string
+	cache   *cache
+}
+
+func (f *factory) Counter(name string, tags map[string]string) metrics.Counter {
+	fullName, fullTags, key := f.getKey(name, tags)
+	c, ok := f.cache.getCounter(key)
```

Comment:
There maybe a race condition here. If 2 threads get here at the same time, they both will see nil counter and then both proceed to create a counter and only one is saved in the cache.

Reply:
Indeed, good catch. We could do a getOrSet() passing a lambda to create if not found. Certainly for metrics related to service names we could easily see a race condition.

Reply:
Why not do what Isaac recommended in other PRs and make the cache a lock free data structure and have the factories, etc. do the locking? This isn't that performance intensive anyway.

Reply:
if the cache is lock-free, then the locking has to be in the `factory`, but `factory` gets recreated every time you call `factory.Namespace()`, and they all share the same cache. I think making cache thread-safe gives better encapsulation in this case.